crates/mpc-protocols/src/mpc_service.rs

```diff
@@ -0,0 +1,115 @@
+/// We want to keep it for some time so leftover messages related to it won't be treated as malicious.
+#[derive(Clone, Copy)]
+enum MPCStatus {
+    Active,
+    Pending,
+    Finished,
+}
+
+/// The `MPCService` is responsible for managing MPC instances:
+/// - It keeps track of all MPC instances
+/// - Runs the MPC session for each active instance
+/// - Ensures that the number of active sessions does not go over `MAX_ACTIVE_MPC_INSTANCES` at the same time
+pub struct MPCService {
+    mpc_instances: HashMap<ObjectID, MPCInstance>,
+    pending: VecDeque<ObjectID>,
+    active_instances_counter: usize,
+}
+
+impl MPCService {
+    pub fn new() -> Self {
+        Self {
+            mpc_instances: HashMap::new(),
+            pending: VecDeque::new(),
+            active_instances_counter: 0,
+        }
+    }
+
+    /// Spawns an asynchronous task to handle incoming messages for a new MPC instance.
+    /// The [`MPCService`] will forward any message related to that instance to this channel.
+    fn spawn_mpc_messages_handler(&self, mut receiver: mpsc::Receiver<MPCInput>) {
+        tokio::spawn(async move {
+            while let Some(message) = receiver.recv().await {
+                // TODO (#235): Implement MPC messages handling
+            }
+        });
+    }
+
+    /// Filter the relevant MPC events from the transaction events & handle them
+    pub fn handle_mpc_events(&mut self, events: &Vec<Event>) -> anyhow::Result<()> {
+        for event in events {
+            if CreatedProofMPCEvent::type_() == event.type_ {
+                let deserialized_event: CreatedProofMPCEvent = bcs::from_bytes(&event.contents)?;
+                self.handle_proof_init_event(deserialized_event);
+                println!("event: CreatedProofMPCEvent {:?}", event);
+            };
+        }
+        Ok(())
+    }
+
+    /// Handles a proof initialization event
+    /// Spawns a new MPC instance if the number of active instances is below the limit
+    /// Otherwise, adds the instance to the pending queue
+    fn handle_proof_init_event(&mut self, event: CreatedProofMPCEvent) {
+        println!(
+            "Received start flow event for session ID {:?}",
+            event.session_id
+        );
+        // let active_instances_counter = self.active_instances_counter.clone().into_inner();
+        // If the number of active instances exceeds the limit, add to pending
+        if self.active_instances_counter >= MAX_ACTIVE_MPC_INSTANCES {
+            self.mpc_instances.insert(
+                event.session_id.clone().bytes,
+                MPCInstance {
+                    status: MPCStatus::Pending,
+                    messages_handler_sender: None,
+                    pending_messages: vec![],
+                },
+            );
+            self.pending.push_back(event.session_id.bytes);
+            return;
+        }
+        let (messages_handler_sender, messages_handler_receiver) = mpsc::channel(100);
```

Comment:
Alert: hard coding `100`.

Reply:
1. Reasonable or not, I'm allergic to 99% of hardcoding (outside of tests) :wink: It obscures behavior, which makes the code hard to understand. By giving a constant a name (e.g., by extracting it into a constant), the reader can realize/double-check _why_ this constant is used and verify it.
2. Can we perhaps derive this from existing constants? For example, does a constant exist that captures the number of active validators? We know that the number of messages to an instance should be, at most, that number multiplied by the maximum number of rounds for a protocol.
3. To close: I don't think 100 is reasonable. If you want to scale this project beyond 100 validators, you'll likely overflow this channel from time to time.

Reply:
Your'e right; I created an issue and added a TODO. The reason I didn't fix it right now is that we change the initialization place of the mpc service in a later PR, and the constructor already receives the number of parties.